lib/net/ssh/known_hosts/cached.rb

```diff
@@ -0,0 +1,127 @@
+require 'digest/md5'
+
+module Net
+  module SSH
+    class KnownHosts
+      # Searches an OpenSSH-style known-host file for a given host, and returns all
+      # matching keys. This is used to implement host-key verification, as well as
+      # to determine what key a user prefers to use for a given host. This is different
+      # From the KnownHosts class in that it builds a cache of parsed known hosts in order
+      # to avoid re-parsing known hosts earch time it is needed.
+      #
+      # An instance of this class can optionaly be used by setting the :known_hosts option.
+      class Cached
+        def initialize(options)
+          @user_files = Array(options[:user_known_hosts_file] || %w[~/.ssh/known_hosts ~/.ssh/known_hosts2])
+          @global_files = Array(options[:global_known_hosts_file] || %w[/etc/ssh/ssh_known_hosts /etc/ssh/ssh_known_hosts2])
+          @md5_sums = {}
+        end
+
+        def search_for(host, options={})
+          # Ensure cache is up to date
+          build_cache if cache_invalid?
+          hostname, host_ip = host.split(',')
+
+          entries = []
+          entries.concat @host_lookups.fetch(hostname, [])
+          entries.concat @pattern_lookups.select { |pattern, _| hostname.match(pattern) }.values.flatten
+          @hmac_lookups.each do |(hmac, salt), entry|
+            digest = OpenSSL::Digest.new('sha1')
+            host_hmac = OpenSSL::HMAC.digest(digest, salt, hostname)
+            entries.concat(entry) if Base64.encode64(host_hmac).chomp == hmac
+          end
+
+          entries.select! { |entry| entry.hosts.include?(host_ip) } if options[:check_host_ip] && host_ip
+
+          HostKeys.new(entries.map(&:key), host, self, options)
+        end
+
+        def build_cache
+          @host_lookups = {}
+          @hmac_lookups = {}
+          @pattern_lookups = {}
+          (@user_files + @global_files).each do |file|
+            parse_known_hosts(File.expand_path(file))
+          end
+        end
+
+        def parse_known_hosts(source)
+          return unless File.readable?(source)
+
+          File.open(source) do |file|
+            @md5_sums[source] = Digest::MD5.hexdigest(File.read(file))
```

Comment:
Please use SHA256 instead of MD5.

Reply:
This seems like an easy change to make. For now i'd like to see what folks think about using a hash instead of mtime. If we want to move forward with hash I think using sha256 should be just fine.

Reply:
Updated to used sha256